src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py

```diff
@@ -58,28 +58,35 @@ def format(self, result):
         Format the query results into the proper dictionary expected at the upper layer Python code.
         The input should be a list of database objects each representing a line returned from the database
         with key names matching the column names from the sql query
-        The result should be a list of dictionaries one record per line returned from the database
-        with key names mapped to the python code expected structures.
+        The result should be a list of dictionaries one record per block returned from the database
```

Comment:
The result is a dictionary of dictionaries, and having the block names as kyes. But that's minor.

Reply:
Did you benchmark it? Having dict of dicts with dynamic keys usually leads to different problems. For instance, in order to parse it where all keys are unknown a-priory the parser code should go through all nested structs. The nested structures usually means that no streaming can be done which again is architectural choice but you should understand the consequences of such choice for your application. This may lead to large memory footprint in application level. By benchmark I mean try to create largest object with this data struct and write code which will parse it. Measure time, and CPU/RAM usage to have clear idea that usage of such data structure will not surprise you in production application.

Reply:
Hi @vkuznet Thanks for mentioning that. This was exactly the reason why at the beginning I opposed the aproach in which the DAO is to do these aggregations in a dictionary form. I just did not expand on the reasons why. And thanks for pointing it out here. These are indeed valid concerns, but we had to compromise. The dynamic keys are needed because otherwise the whole logic for how those SQL queries are constructed and how the checks for deletable blocks are preformed need to be changed significantly. Which would be extremely dangerous now as we are in the middle of the data taking. 
Benchmark and scalability tests and memory footprint comparisons between the two DAO outputs are not done - this I am sure about. But we also very well know that no additional parsing of the returned object is done rather than a single iteration and a set comparison for all of its elements is happening. Which should not increase the code complexity more than to a O(n).  As of memory foot print, even if a single DAO output does increase in size due to the data structure change....well this one is used only once during the polling cycle and immediately gets out of scope after the method is executed, so again I wouldn't expect more than a constant increase in that regard either. And here are the reasons why I did agree with the so suggested change of the data structure returned by the DAO. 

And just a hint - the code is already applied through a patch in production so we are feeling quite confident it is working at least well enough for the current load at T0.

Reply:
Todor, with all my respect, this does not answer my questions. I hope you also understand that by providing these  "explanations" we only hide the issue or delegate it to another part of infrastructure. When you claim that code in production and it runs, does it mean that you measure the impact of this change? What is it? Does this change RAM by 1% or double it. We may still run within the resource allocation but we may enter dangerous zone or thing may break if object will get larger. Let's be professional here. Please estimate the impact and provide concrete numbers. Without them any explanation does not make any sense to me but its impact may not be immediate but still valid and dangerous, especially during data taking.

Reply:
While not a proper benchmark, I took a few points of data on the performance before and after the patch. These data points are taken with 9796 blocks in the system, a normal number after a few weeks of heavy data taking (we peaked at around 13k blocks during July). 

**Without patch**
These 9796 blocks translates into a list of 40665 elements. 
PSS increased around 50MB after calling the query
RucioInjectorPoller took around 30-40 secs to execute

**With patch**
These 9796 blocks translates into a dict of 9796 elements. 
PSS increased around 80MB after calling the query
RucioInjectorPoller took around 15-25 secs to execute

Reply:
Thanks @germanfgv for doing the check. This completely overlaps with what I was expecting and was mentioning above. To me it seems pretty affordable increase of resource consumption, especially when it comes to memory. And moreover we are actually performing better in terms of iterating trough the data structure in the component (outside the DAO), as seen from your tests, because the data is already pre-aggregated. 
@vkuznet  I hope these numbers are good enough for you as well.

Reply:
thanks @germanfgv and @todor-ivanov, this is certainly a very assuring observation and I'm fine to proceed with a patch.

Reply:
Thanks @vkuznet! I'll proceed with merging now.